lib/pdu.c

```diff
@@ -397,15 +397,23 @@ smb2_find_pdu(struct smb2_context *smb2,
 static int
 smb2_is_error_response(struct smb2_context *smb2,
                        struct smb2_pdu *pdu) {
-        if ((smb2->hdr.status & SMB2_STATUS_SEVERITY_MASK) ==
+        if ((smb2->hdr.status & SMB2_STATUS_ERROR_SEVERITY_MASK) ==
             SMB2_STATUS_SEVERITY_ERROR) {
                 switch (smb2->hdr.status) {
                 case SMB2_STATUS_MORE_PROCESSING_REQUIRED:
                         return 0;
                 default:
                         return 1;
                 }
-        }
+        } else if ((smb2->hdr.status & SMB2_STATUS_WARNING_SEVERITY_MASK) ==
```

Comment:
-- not sure if this is the best way to handle this. Please suggest an alternative.
-- I am not familiar with samba code base so not sure how it is handled there.
-- also, there are many other warning codes from SMB specs which are not handled here. For now, I have just taken the conservative approach: kept the existing behaviour for all warning codes except the one I faced a problem with.

Reply:
This should just be } else if ((smb2->hdr.status & SMB2_STATUS_SEVERITY_MASK) ==

Here is why this would be wrong.
Imagine you have a status 0xc0000016  which is an error, not a warning.

IF you only check this mask using a single bit for the bitmask, i.e. SMB2_STATUS_WARNING_SEVERITY_MASK==0x80000000

Then this code becomesITY_MASK
if ((status & SMB2_STATUS_WARNING_SEVERITY_MASK) == SMB2_STATUS_SEVERITY_WARNING)
which would be

But status/0xc0000016 masked with SMB2_STATUS_WARNING_SEVERITY_MASK/0x8000000
which is equal to SMB2_STATUS_SEVERITY_WARNING

I.e.   0xc0000016 is an error, not a warning, but this if statement would still be true for errors.